[Docs] add practical example to improve data management doc (#5844)
flyteorg/flyte: docs/user_guide/concepts/main_concepts/data_management.rst

Types of Data
=============

In Flyte, data is categorized into metadata and raw data to optimize data handling and improve performance and security.

> * **Metadata**: Small values, like integers and strings, are treated as "stack parameters" (passed by value). This metadata is globally accessible to Flyte components (FlytePropeller, FlyteAdmin, and other running pods/jobs). Each entry is limited to 10MB and is passed directly between tasks. On top of that, metadata allow in-memory computations for branches, partial outputs, and composition of multiple outputs as input for other tasks.

Reviewer: > `Each entry is limited to 10MB and is passed directly between tasks.`

For future proofing of this section, shouldn't we add here that bigger files are automatically offloaded?
cc @pingsutw

    Tasks can then access this data via Flyte’s automated download or streaming, enabling efficient access to large datasets without needing to transfer full copies.

*Source code reference for auto-offloading value sizes limitation:*

.. literalinclude:: https://github.com/flyteorg/static-resources/blob/ddec386ab74897e3c09d89fdaf305e77a2ecc52c/flyte/flytepropeller/pkg/controller/config/config.go
   :caption: flytepropeller/pkg/controller/config/config.go
   :language: go
   :lines: 184-192

Data Flow and Security
~~~~~~~~~~~~~~~~~~~~~~

Flyte’s data separation avoids bottlenecks and security risks:

* **Metadata** remains within Flyte’s control plane, making it accessible through the Flyte Console or CLI.
* **Raw Data** is accessible only by tasks, stored securely in an external blob store, preventing Flyte’s control plane from directly handling large data files.

Moreover, a unique property of this separation is that all meta values are read by FlytePropeller engine and available on the FlyteConsole or CLI from the control plane.

Example
~~~~~~~

Consider a basic Flyte task:

> .. code-block:: python
>     @task

Reviewer 2: ```suggestion
.. code-block:: python

    @task
```

Types of Data
=============

In Flyte, data is categorized into metadata and raw data to optimize data handling and improve performance and security.

> * **Metadata**: Small values, like integers and strings, are treated as "stack parameters" (passed by value). This metadata is globally accessible to Flyte components (FlytePropeller, FlyteAdmin, and other running pods/jobs). Each entry is limited to 10MB and is passed directly between tasks. On top of that, Metadata allow in-memory computations for branches, partial outputs, and composition of multiple outputs as input for other tasks.

Author: Okay, fixed.

* **Raw data**: Larger data, such as files and dataframes, are treated as "heap parameters" (passed by reference). Flyte stores raw data in an object store (e.g., S3), uploading it on first use and passing only a reference thereafter. Tasks can then access this data via Flyte’s automated download or streaming, enabling efficient access to large datasets without needing to transfer full copies.

*Source code reference for auto-offloading value sizes limitation:*

.. raw:: html

>    <a href="https://github.com/flyteorg/flyte/blob/6c4f8dbfc6d23a0cd7bf81480856e9ae1dfa1b27/flytepropeller/pkg/controller/config/config.go#L184-L192">View source code on GitHub</a>

Author: @Future-Outlier Since `.rst` does not support remote link references, I am using HTML to force the use of the gitsha link.

   :lines: 184-192

Data Flow and Security
~~~~~~~~~~~~~~~~~~~~~~

Flyte’s data separation avoids bottlenecks and security risks:

* **Metadata** remains within Flyte’s control plane, making it accessible through the Flyte Console or CLI.

> * **Raw Data** is accessible only by tasks, stored securely in an external blob store, preventing Flyte’s control plane from directly handling large data files.

Reviewer 3: ```suggestion
* **Raw data** is accessible only by tasks, stored securely in an external blob store, preventing Flyte’s control plane from directly handling large data files.
```

Author: Okay, fixed.

Moreover, a unique property of this separation is that all meta values are read by FlytePropeller engine and available on the FlyteConsole or CLI from the control plane.

Example
~~~~~~~

Consider a basic Flyte task:

.. code-block:: python
    @task
    def my_task(m: int, n: str, o: FlyteFile) -> pd.DataFrame:
       ...

In this task, ``m``, ``n``, and ``o`` are inputs: ``m`` (int) and ``n`` (str) are simple types, while ``o`` is a large, arbitrarily sized file.
Flyte treats each differently:

* Metadata: Small values like ``m`` and ``n`` are inlined within Flyte’s metadata and passed directly between tasks.

> * Raw Data: Objects like ``o`` and the output pd.DataFrame are offloaded to an object store (e.g., S3), with only references retained in metadata.

Reviewer 3: ```suggestion
* Raw data: Objects like ``o`` and the output ``pd.DataFrame`` are offloaded to an object store (e.g., S3), with only references retained in metadata.
```

Author: Okay, fixed.

> Flytekit TypeTransformers make it possible to use complex objects as if they are available locally - just like persistent filehandles. But Flyte backend only deals with the references.

Reviewer 3: ```suggestion
Flytekit TypeTransformers make it possible to use complex objects as if they are available locally, just like persistent filehandles. However, the Flyte backend only deals with the references.
```

Author: Okay, fixed.

But for Metadata, the data should be accessible to Flyte control plane.

Data persistence is also pluggable. By default, it supports all major blob stores and uses an interface defined in Flytestdlib.

Practical Example
~~~~~~~~~~~~~~~~~

Let's consider a simple example where we have some tasks that needs to operate huge dataframes.

> The first task reads a file from the object store, shuffles the data, save to local disk and passes the path to the next task.

Reviewer 3: ```suggestion
The first task reads a file from the object store, shuffles the data, saves to local disk, and passes the path to the next task.
```

Author: Okay, fixed.

.. code-block:: python

    @task()
    def task_remove_column(input_file: FlyteFile, column_name: str) -> FlyteFile:
        """
        Reads the input file as a DataFrame, removes a specified column, and outputs it as a new file.
        """
        input_file.download()
        df = pd.read_csv(input_file.path)

        # remove column
        if column_name in df.columns:
            df = df.drop(columns=[column_name])

        output_file_path = "data_finished.csv"
        df.to_csv(output_file_path, index=False)

        return FlyteFile(output_file_path)
       ...

> The second task reads the file from the previous task, removes a column, save to local disk and return the path.

Reviewer 3: ```suggestion
The second task reads the file from the previous task, removes a column, saves to local disk, and returns the path.
```

Author: Okay, fixed.

.. code-block:: python

    @task()
    def task_remove_column(input_file: FlyteFile, column_name: str) -> FlyteFile:
        """
        Reads the input file as a DataFrame, removes a specified column, and outputs it as a new file.
        """
        input_file.download()
        df = pd.read_csv(input_file.path)

        # remove column
        if column_name in df.columns:
            df = df.drop(columns=[column_name])

        output_file_path = "data_finished.csv"
        df.to_csv(output_file_path, index=False)

        return FlyteFile(output_file_path)
       ...

> And this is how the workflow looks like:

Reviewer 3: ```suggestion
And here is the workflow:
```

Author: Okay, fixed.

.. code-block:: python
    @task()
    def task_remove_column(input_file: FlyteFile, column_name: str) -> FlyteFile:
        """
        Reads the input file as a DataFrame, removes a specified column, and outputs it as a new file.
        """
        input_file.download()
        df = pd.read_csv(input_file.path)

        # remove column
        if column_name in df.columns:
            df = df.drop(columns=[column_name])

        output_file_path = "data_finished.csv"
        df.to_csv(output_file_path, index=False)

        return FlyteFile(output_file_path)
       ...

And this is how the workflow looks like:

.. code-block:: python
    @workflow
    def wf() -> FlyteFile:
        # In Flyte sandbox, existing files must be accessed within the workflow, not directly in a task.
        existed_file = FlyteFile("s3://custom-bucket/data.csv")
        shuffled_file = task_read_and_shuffle_file(input_file=existed_file)
        result_file = task_remove_column(input_file=shuffled_file, column_name="County")
        return result_file
       ...

This example shows how to access an existing file in a MinIO bucket from the Flyte Sandbox and pass it between tasks with ``FlyteFile``.

> * Files in MinIO must be accessed at the workflow level in Flyte Sandbox, then passed to tasks to avoid permission errors.

Reviewer 2: Only the task will access the file

But for Metadata, the data should be accessible to Flyte control plane.

Data persistence is also pluggable. By default, it supports all major blob stores and uses an interface defined in Flytestdlib.

Example
~~~~~~~

Let's consider a simple example where we have some tasks that needs to operate huge dataframes.

The first task reads a file from the object store, shuffles the data, save to local disk and passes the path to the next task.

.. code-block:: python
    @task()
    def task_remove_column(input_file: FlyteFile, column_name: str) -> FlyteFile:

>         """

Reviewer 4: Not sure is this the right way or should we move to flytesnacks

        return FlyteFile(output_file_path)
       ...

And this is how the workflow looks like:

.. code-block:: python
    @workflow
    def wf() -> FlyteFile:
        # In Flyte sandbox, existing files must be accessed within the workflow, not directly in a task.
        existed_file = FlyteFile("s3://custom-bucket/data.csv")
        shuffled_file = task_read_and_shuffle_file(input_file=existed_file)
        result_file = task_remove_column(input_file=shuffled_file, column_name="County")
        return result_file
       ...

This example shows how to access an existing file in a MinIO bucket from the Flyte Sandbox and pass it between tasks with ``FlyteFile``.

* Files in MinIO must be accessed at the workflow level in Flyte Sandbox, then passed to tasks to avoid permission errors.
* When a workflow outputs a local file as a ``FlyteFile``, Flyte automatically uploads it to MinIO and provides an S3 URL for downstream tasks, no manual uploads needed. Take a look at the following:

First task output metadata:

.. image:: https://raw.githubusercontent.com/flyteorg/static-resources/main/flyte/concepts/data_movement/flyte_data_movement_example_output.png

Second task input metadata:

> .. image:: https://raw.githubusercontent.com/flyteorg/static-resources/main/flyte/concepts/data_movement/flyte_data_movement_example_input.png

Reviewer 4: use gitsha first, then change it back.
gitsha: 
https://github.com/flyteorg/static-resources/blob/ddec386ab74897e3c09d89fdaf305e77a2ecc52c/flyte/concepts/data_movement/flyte_data_movement_example_input.png

Author: 719db18af583f727d5b3c099e01d031a22a3216b fixed
